dolphin/alert_manager/trap_receiver.py

```diff
@@ -0,0 +1,137 @@
+            if load_mibs_list:
+                _mibs = LOAD_MIB_MODULE.split(",")
+                mibBuilder.loadModules(*_mibs)
+        except error.MibNotFoundError as excep:
+            LOG.info(" {} Mib Not Found!".format(excep))
+
+    def _add_transport(self):
+        """
+        :return:
+        """
+        try:
+            config.addTransport(
+                self.snmpEngine,
+                udp.domainName,
+                udp.UdpTransport().openServerMode((constants.DEF_TRAP_RECV_ADDR, int(constants.DEF_TRAP_RECV_PORT)))
+            )
+        except Exception as e:
+            LOG.info("{} Port Binding Failed the Provided Port {} is in Use".format(e, self.receiverPort))
+
+    def _cbFun(self, stateReference, contextEngineId, contextName,
+              varBinds, cbCtx):
+        global mibViewController
+        LOG.info("####################### NEW Notification #######################")
+        execContext = self.snmpEngine.observer.getExecutionContext('rfc3412.receiveMessage:request')
+        LOG.info(
+            '#Notification from %s \n#ContextEngineId: "%s" \n#ContextName: "%s" \n#SNMPVER "%s" \n#SecurityName "%s"' % (
+            '@'.join([str(x) for x in execContext['transportAddress']]), contextEngineId.prettyPrint(),
+            contextName.prettyPrint(), execContext['securityModel'], execContext['securityName']))
+        for oid, val in varBinds:
+            output = rfc1902.ObjectType(rfc1902.ObjectIdentity(oid), val).resolveWithMib(
+                mibViewController).prettyPrint()
+            LOG.info(output)
+
+    def _configure_snmp_userpara(self):
+        """
+        :return:
+        """
+
+        COMMUNITYSTRING = constants.SNMP_COMMUNITY_STR
+        config.addV1System(self.snmpEngine, COMMUNITYSTRING, COMMUNITYSTRING)
+        __authProtocol = {
+            'usmHMACMD5AuthProtocol': config.usmHMACMD5AuthProtocol,
+            'usmHMACSHAAuthProtocol': config.usmHMACSHAAuthProtocol,
+            'usmAesCfb128Protocol': config.usmAesCfb128Protocol,
+            'usmAesCfb256Protocol': config.usmAesCfb256Protocol,
+            'usmAesCfb192Protocol': config.usmAesCfb192Protocol,
+            'usmDESPrivProtocol': config.usmDESPrivProtocol,
+            'usmNoAuthProtocol': config.usmNoAuthProtocol,
+            'usmNoPrivProtocol': config.usmNoPrivProtocol
+        }
+
+
+        config.addV3User(
+            self.snmpEngine, userName=constants.SNMP_USM_USER,
+            authKey=constants.SNMP_V3_AUTHKEY, privKey=constants.SNMP_V3_PRIVKEY,
+            authProtocol=__authProtocol.get(
+                constants.SNMP_V3_AUTH_PROTOCOL, config.usmNoAuthProtocol),
+            privProtocol=__authProtocol.get(
+                constants.SNMP_V3_PRIV_PROTOCOL, config.usmNoPrivProtocol),
+            securityEngineId=v2c.OctetString(
+                hexValue=constants.SNMP_ENGINE_ID))
+
+        return
+
+    def start_trap_receiver(self):
+        snmpEngine = engine.SnmpEngine()
```

Comment:
engine is the the device who send trap , right?

Reply:
engine is like a server context at receiver side, everything is attached to this engine